### PR TITLE
[Merged by Bors] - feat(Data/ENNReal/Real): add lemmas about iInf, iSup

### PR DESCRIPTION
**Pull request by mariainesdff**
Used in #15373.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/ENNReal/Real.lean [hunk 1]
```diff
@@ -571,6 +571,47 @@ See also `ENNReal.mul_iInf` that assumes `[Nonempty ι]` but does not require `x
...
+theorem le_iSup_prop (u : ℕ → ℝ≥0∞) {n k : ℕ} (hnk : n ≤ k) : u k ≤ ⨆ (k : ℕ) (_ : n ≤ k), u k := by
+  apply le_iSup_of_le k
+  rw [ciSup_pos hnk]
+
+/-- The function sending `n : ℕ` to `⨆ (k : ℕ) (x : n ≤ k), u k` is antitone. -/
+theorem Antitone.iSup {u : ℕ → ℝ≥0∞} : Antitone fun n : ℕ ↦ ⨆ (k : ℕ) (_ : n ≤ k), u k :=
+  antitone_nat_of_succ_le (fun n ↦ iSup₂_le_iff.mpr
+    fun _ hk ↦ le_iSup_prop u (le_trans (Nat.le_succ n) hk))
+
+/-- If `u : ℕ → ℝ≥0∞` is bounded above by a real number, then its `supr` is finite. -/
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/ENNReal/Real.lean:607]: We no longer have `supr`

File: Mathlib/Data/ENNReal/Real.lean [hunk 2]
```diff
@@ -571,6 +571,47 @@ See also `ENNReal.mul_iInf` that assumes `[Nonempty ι]` but does not require `x
...
+theorem iSup_le_top_of_bddAbove {u : ℕ → ℝ≥0∞} {B : ℝ≥0} (hu : ∀ x, u x ≤ B) (n : ℕ) :
+    (⨆ (k : ℕ) (_ : n ≤ k), u k) ≠ ⊤ :=
+  haveI h_le : (⨆ (k : ℕ) (_ : n ≤ k), u k) ≤ B := by
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/ENNReal/Real.lean:610]:
```suggestion
  have h_le : (⨆ (k : ℕ) (_ : n ≤ k), u k) ≤ B := by
```

File: Mathlib/Data/ENNReal/Real.lean [hunk 3]
```diff
@@ -571,6 +571,47 @@ See also `ENNReal.mul_iInf` that assumes `[Nonempty ι]` but does not require `x
...
+theorem iInf_mul_le_mul_iInf {u v : ℕ → ℝ≥0∞} (hu_top : ∀ x, u x ≠ ⊤) (hu : Antitone u)
+    (hv_top : ∀ x, v x ≠ ⊤) (hv : Antitone v) : iInf (u * v) ≤ iInf u * iInf v := by
+  rw [iInf_le_iff]
+  intro b hb
+  apply le_iInf_mul_iInf hu_top hv_top
+  intro m n
+  exact le_trans (hb (max m n)) (mul_le_mul' (hu (le_max_left _ _)) (hv (le_max_right _ _)))
+
+theorem iSup_tail_seq (u : ℕ → ℝ≥0∞) (n : ℕ) :
+    (⨆ (k : ℕ) (_ : n ≤ k), u k) = ⨆ k : { k : ℕ // n ≤ k }, u k := by rw [iSup_subtype]
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/ENNReal/Real.lean:596]: What's the advantage over `iSup_subtype` directly?

**Comment by D-Thomine** [comment]
In `le_iInf_mul_iInf` and `iInf_mul_le_mul_iInf`, it seems that the hypotheses `(hu_top : ∀ x, u x ≠ ⊤)` and `(hv_top : ∀ x, v x ≠ ⊤)` could use existential quantifiers instead of universal quantifiers. Basically, the idea is that you want to avoid e.g. `f` infinite everywhere and `g` converging to 0, in which case `H` is vacuous but the conclusion is too strong. But existential quantifiers are enough to guarantee that neither `f` nor `g` are infinite everywhere.

**Comment by D-Thomine** [comment]
On a side note: some time ago, we added things like `[EReal.add_iInf_le_iInf_add](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Instances/EReal.html#EReal.add_iInf_le_iInf_add)`. Given that `(Ereal, +)` is isomorphic to `(ENNReal, *)` through the exponential, you may easily get some similar statements.

File: Mathlib/Data/ENNReal/Real.lean (hunk 3 above)
Comment (mariainesdff) [reply Mathlib/Data/ENNReal/Real.lean:596]: I have removed this and inlined it in the proof of `iSup_le_top_of_bddAbove`.

**Comment by mariainesdff** [comment]
> Is this something you could prove instead?

This version is sufficient for my use case, but if you have formalized the more general version I would have no problem changing it.

File: Mathlib/Data/ENNReal/Real.lean [hunk 4]
```diff
@@ -558,6 +558,28 @@ theorem iInf_sum {α : Type*} {f : ι → α → ℝ≥0∞} {s : Finset α} [No
...
+-- (Yaël) This theorem is too specialised
+/-- The function sending `n : ℕ` to `⨆ (k : ℕ) (x : n ≤ k), u k` is antitone. -/
+theorem Antitone.iSup {u : ℕ → ℝ≥0∞} : Antitone fun n : ℕ ↦ ⨆ (k : ℕ) (_ : n ≤ k), u k :=
+  fun _m _n hmn ↦ biSup_mono fun _ ↦ hmn.trans
+
+-- (Yaël) This theorem is too specialised
+/-- If `u : ℕ → ℝ≥0∞` is bounded above by a real number, then its `iSup` is finite. -/
+theorem iSup_ne_top_of_bddAbove {u : ℕ → ℝ≥0∞} {B : ℝ≥0} (hu : ∀ x, u x ≤ B) (n : ℕ) :
+    (⨆ (k : ℕ) (_ : n ≤ k), u k) ≠ ⊤ :=
+  ne_top_of_le_ne_top coe_ne_top <| iSup₂_le fun m _ ↦ hu m
```
Comment (YaelDillies) [inline Mathlib/Data/ENNReal/Real.lean:581]: As written in the comments, I think all those lemmas should be inlined

**Review by YaelDillies — APPROVED** [review]
LGTM but I have written most of the current content of the PR

maintainer merge

**Comment by urkud** [comment]
Thanks! :tada:
bors merge

**Pull request closed without merging**